<launch>
  <arg name="agent_id" default= "0" />
  <arg name="scale"    default="0.5" />
  <arg name="process_every_nth_frame" default="3" />
  
  <!-- PCL Fusion Node -->
  <node name="pcl_fusion_$(arg agent_id)" pkg="pcl_fusion" type="pcl_fusion_node" output="screen">
    <param name="cam_config" type="string" value="$(find pcl_fusion)/conf/ncamera.yaml"/>
    <param name="kf_per_cloud"         value="8"/>
    <param name="voxel_filter_size"    value="0.07"/>
    <param name="agent_id"             value="$(arg agent_id)"/>
    <param name="num_odom_connections" value="2"/>

    <remap from="/pointcloud"    to="/dense_stereo/pointcloud" />
    <remap from="/vins_keyframe" to="/pose_graph_$(arg agent_id)/vins_keyframe" /> 
  </node>

  <!-- Dense stereo Node -->
  <node name="dense_stereo" pkg="image_undistort" type="dense_stereo_node" output="log">
    <param name="input_camera_info_from_ros_params" value="true"/>
    <param name="first_camera_namespace"            value="cam0"/>
    <param name="second_camera_namespace"           value="cam1"/>
    <param name="first_output_frame"                value="cam0_rect"/>
    <param name="second_output_frame"               value="cam1_rect"/>
    <param name="scale"                             value="$(arg scale)"/>
    <param name="depth/use_sgbm"                    value="true"/>
    <param name="depth/do_median_blur"              value="false"/>
    <param name="depth/use_mode_HH"                 value="true"/>
    <param name="process_every_nth_frame"           value="$(arg process_every_nth_frame)"/>
    <rosparam file="$(find pcl_fusion)/conf/euroc_camchain.yaml"/>

    <remap from="raw/first/image"        to="cam0_$(arg agent_id)/image_raw"/>
    <remap from="raw/second/image"       to="cam1_$(arg agent_id)/image_raw"/>
    <remap from="raw/first/camera_info"  to="cam0/camera_info"/>
    <remap from="raw/second/camera_info" to="cam1/camera_info"/>
  </node>
 
</launch>
